metaflow/plugins/frameworks/checkpoints.py

```diff
@@ -0,0 +1,137 @@
+import logging
+import os
+from typing import List, NamedTuple, Optional, TypeVar, Union
+from urllib.parse import urlparse
+
+from metaflow.current import current
+from metaflow.datatools import S3
+from metaflow.flowspec import FlowSpec
+from metaflow.client.core import Run
+from metaflow.datatools.s3 import S3Object
+
+_logger = logging.getLogger(__name__)
+_logger.addHandler(logging.StreamHandler())
+_logger.setLevel(logging.INFO)
+
+T = TypeVar("T")
+
+
+def _get_s3_latest_checkpoint_s3_info(s3_checkpoint_root: str) -> Optional[S3Object]:
+    s3 = S3(s3root=s3_checkpoint_root)
+    checkpoints = s3.list_paths([""])
+    if not checkpoints:
+        return None
+
+    with_infos: List = s3.info_many([cpt.key for cpt in checkpoints])
+    return max(with_infos, key=lambda info: info.last_modified)
+
+
+def _get_latest_checkpoint_name(root: str) -> str:
+    if urlparse(root).scheme == "s3":
+        latest = _get_s3_latest_checkpoint_s3_info(root)
+        return latest.key if latest else None
+    else:
+        files = os.listdir(root)
+        paths = [os.path.join(root, basename) for basename in files]
+        paths = [path for path in paths if os.path.isfile(path)]
+        return max(paths, key=os.path.getctime)
+
+
+def _get_checkpoint_root(
+    run: Optional[Union[FlowSpec, Run]] = None,
+    step_name: Optional[str] = None,
+    task_id: Optional[str] = None,
+) -> str:
+    temp_step_name = step_name if step_name else current.step_name
+    temp_task_id = task_id if task_id else current.task_id
+    return os.path.join(
+        _get_run_root(run), f"checkpoints/{temp_step_name}/{temp_task_id}"
+    )
+
+
+def _get_run_root(run: Optional[Union[FlowSpec, Run]] = None) -> str:
+    """Function separated out to unit test mock"""
+    root_env = os.environ.get("CHECKPOINT_ROOT")
+    if root_env:
+        return root_env
+    else:
+        temp_run = run if run else current.flow
+        return S3(run=temp_run)._s3root
+
+
+def _get_resume_checkpoint_path(
+    root: str,
+    resume_checkpoint_path: Optional[str] = None,
+) -> Optional[str]:
+    if current.retry_count == 0:
+        return resume_checkpoint_path
+    elif current.retry_count > 0:
+        key = _get_latest_checkpoint_name(root)
+        if key:
+            ret = os.path.join(root, key)
+            _logger.debug(f"_get_resume_checkpoint_path returning {ret}")
+            return ret
+        else:
+            _logger.info(
+                f"{current.retry_count=} but using {resume_checkpoint_path=} because no checkpiont found."
+            )
+            return resume_checkpoint_path
+
+
+CheckpointPaths = NamedTuple(
+    "CheckpointPaths", [("root", str), ("resume_path", Optional[str])]
+)
+
+
+def get_checkpoint_paths(
+    resume_checkpoint_path: Optional[str] = None,
+) -> CheckpointPaths:
+    """
+    This function gets the checkpoint root and resume path for a Flow Run step.
```

Comment:
It might be nice to discuss how checkpointing allows a flow to use spot or handle losing a node i.e. what is checkpointing used for.

Reply:
done!